File: app/src/main/java/com/chooloo/www/callmanager/ui/cursor/CursorAdapter.java
package com.chooloo.www.callmanager.ui.cursor;

import android.content.Context;
import android.database.Cursor;
import android.database.DataSetObserver;
import android.util.ArrayMap;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import timber.log.Timber;

public abstract class CursorAdapter<VH extends RecyclerView.ViewHolder> extends RecyclerView.Adapter<VH> {

    protected Cursor mCursor;
    protected Context mContext;
    protected DataSetObserver mDataSetObserver;
    protected ArrayMap<VH, Integer> mViewHoldersMap;

    public CursorAdapter(Context context) {
        mContext = context;
        setUp();
    }

    @Override
    public int getItemCount() {
        return mCursor != null ? mCursor.getCount() : 0;
    }

    @Override
    public void setHasStableIds(boolean hasStableIds) {
        super.setHasStableIds(true);
    }

    protected void setUp() {
        mViewHoldersMap = new ArrayMap<>();

        mDataSetObserver = new DataSetObserver() {
            @Override
            public void onChanged() {
                super.onChanged();
                notifyDataSetChanged();
            }

            @Override
            public void onInvalidated() {
                super.onInvalidated();
                notifyDataSetChanged();
            }
        };
        registerDataSetObserver(mDataSetObserver);
    }

    public void setCursor(Cursor newCursor) {
        if (newCursor != mCursor) {
            unregisterDataSetObserver();
            if (mCursor != null) mCursor.close();
            mCursor = newCursor;
            setUp();
        }
        notifyDataSetChanged();
    }

    public Cursor getCursor() {
        return mCursor;
    }

    public void registerDataSetObserver(DataSetObserver dataSetObserver) {
        if (mCursor != null) {
            mCursor.registerDataSetObserver(dataSetObserver);
        }
    }

    public void unregisterDataSetObserver() {
        if (mCursor != null && mDataSetObserver != null) {
            try {
                mCursor.unregisterDataSetObserver(mDataSetObserver);
            } catch (IllegalStateException e) {
                Timber.e("CursorAdapter was not registered when trying to unregister");
            }
        }
    }
}

File: app/src/main/java/com/chooloo/www/koler/adapter/CursorAdapter.kt
package com.chooloo.www.koler.adapter

import android.content.Context
import android.database.Cursor
import android.database.DataSetObserver
import androidx.recyclerview.widget.RecyclerView
import timber.log.Timber

abstract class CursorAdapter<VH : RecyclerView.ViewHolder?>(
    protected var context: Context
) : RecyclerView.Adapter<VH>() {

    protected var _cursor: Cursor? = null
    protected var _dataSetObserver: DataSetObserver = object : DataSetObserver() {
        override fun onChanged() {
            super.onChanged()
            notifyDataSetChanged()
        }

        override fun onInvalidated() {
            super.onInvalidated()
            notifyDataSetChanged()
        }
    }

    var cursor: Cursor?
        get() = _cursor
        set(value) {
            if (value !== _cursor) {
                if (_cursor != null) {
                    unregisterDataSetObserver(_dataSetObserver)
                    _cursor!!.close()
                }
                _cursor = value
                registerDataSetObserver(_dataSetObserver)
            }
        }

    override fun getItemCount() = _cursor?.count ?: 0

    override fun setHasStableIds(hasStableIds: Boolean) {
        super.setHasStableIds(true)
    }

    private fun registerDataSetObserver(dataSetObserver: DataSetObserver) {
        _cursor?.registerDataSetObserver(dataSetObserver)
    }

    private fun unregisterDataSetObserver(dataSetObserver: DataSetObserver) {
        try {
            _cursor?.unregisterDataSetObserver(dataSetObserver)
        } catch (e: IllegalStateException) {
            Timber.e("CursorAdapter was not registered when trying to unregister")
        }
    }
}